### ISSUE TITLE
False positive: print_literal and env!

### ISSUE THREAD

**Issue by bjgill** [rust-lang-nursery/rust-clippy#2657]
Found with clippy 0.0.193

```
warning: printing a literal with an empty format string
  --> src/bin/rm/main.rs:92:41
   |
92 |         println!("cargo-rm version {}", env!("CARGO_PKG_VERSION"));
   |                                         ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(print_literal)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.193/index.html#print_literal
```

I'm fairly certain that there isn't a way to shorten this as the docs suggest. 

It looks as if this was discussed in https://github.com/rust-lang-nursery/rust-clippy/issues/2617 and #2608. Whilst it's true that `println!(env!("FOO"))` is better than `println!("{}", env!("FOO"))`, I don't think there is an equivalent for `println!("bar: {}", env!("FOO"))`.

**Comment by mrecachinas** [comment]
Yeah this is a tricky one... using a late pass converts all macros, so `env!(“FOO”)` looks like a string by the time the `print_literal` check happens.

**Comment by mrecachinas** [comment]
So I looked a bit more into this. I added a test for `env!("FOO")` where I `export FOO=test`. As expected, the test yielded a false positive:

```
error: printing a literal with an empty format string
  --> tests/ui/print_literal.rs:40:25
   |
40 |     println!("ENV: {}", env!("FOO"));
   |                         ^^^^^^^^^^^
```

I then tried to inspect what's going on as well as use `utils::{in_macro}`. Unfortunately, `in_macro` is `false` for the exact `span` that gets warned in the lint. 

Looking into the args at the time it's linted in `write.rs`, you can see `expr(896: &match (&"test",) ...)`

```
[expr(885: &["ENV: ", "\n"]), expr(896: &match (&"test",) {
     (__arg0,) => [<::fmt::ArgumentV1>::new(__arg0, ::fmt::Display::fmt)],
 }), expr(908: &[::fmt::rt::v1::Argument{position: ::fmt::rt::v1::Position::At(0usize),
                          format:
                              ::fmt::rt::v1::FormatSpec{fill: ' ',
                                                        align:
                                                            ::fmt::rt::v1::Alignment::Unknown,
                                                        flags: 0u32,
                                                        precision:
                                                            ::fmt::rt::v1::Count::Implied,
                                                        width:
                                                            ::fmt::rt::v1::Count::Implied,},}])]
```

I'm very curious what the way to proceed here would be. I'm thinking this might require an EarlyLintPass, but I'm not entirely sure how to chain both together.

**Comment by oli-obk** [comment]
I assumed I fixed this with https://github.com/rust-lang/rust/pull/44248

Maybe it regressed again in rustc?

**Comment by mbr** [comment]
Still present on rustc 1.29.0-nightly (64f7de921 2018-07-12), clipp 0.0.211.

**Comment by phansch** [comment]
I can't reproduce this on the [playground][playground] anymore, so I'm going ahead and will close this issue. Feel free to re-open if this still happens with a recent Clippy version.

[playground]: https://play.rust-lang.org/?gist=c76e0c67e69890b84c271602ec95c0da&version=stable&mode=debug&edition=2015
